### PR TITLE
Add Protobuf Mutator Fuzz Test

### PR DESCRIPTION
**Pull request by br-lewis**
This adds a fuzz test for the protobuf mutator in selffuzz

### CODE REVIEW

File: selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      throws IOException {
+    if (messageType == null) {
+      return;
+    }
+
+    protoName = messageType.getName();
+    // the name has to be valid to create the filedescriptor, other invalid names will be caught
+    // when constructing the mutator
+    Matcher matcher = protoNamePattern.matcher(protoName);
+    if (!matcher.find()) {
```
Comment (fmeum) [inline selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java:57]:
```suggestion
    if (!protoNamePattern.matcher(protoName).matches()) {
```

File: selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    SerializingMutator<DynamicMessage> mutator;
+    try {
+      mutator = (SerializingMutator<DynamicMessage>) ProtoMutators.newFactory().createOrThrow(
+          new TypeHolder<@WithDefaultInstance(
+              "com.code_intelligence.selffuzz.mutation.mutator.proto.ProtobufMutatorFuzzTest#getDefaultInstance")
+              DynamicMessage>() {
+          }.annotatedType());
+    } catch (IllegalArgumentException e) {
+      // an invalid proto descriptor will throw a DescriptorValidationException below but by the
```
Comment (fmeum) [inline selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java:73]: Could you provide me with a stack trace of such an error? If this exception has such a complicated form, we may want to catch this in the mutator factory and provide a more comprehensible error message.

**Review by fmeum — APPROVED** [review]
Looks good, just a few minor comments.

File: selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  private static FileDescriptorProto file;
+
+  // https://protobuf.dev/reference/protobuf/proto3-spec/#identifiers
+  private static final Pattern protoNamePattern = Pattern.compile("^[a-zA-Z][a-zA-Z._]$");
+
+  @SuppressWarnings({"unchecked", "unused"})
+  @FuzzTest
+  void protobufMutatorTest(long seed, @NotNull DescriptorProto messageType, byte @NotNull[] bytes)
+      throws IOException {
+    if (messageType == null) {
```
[the remark this answers is not in the document]
Comment (br-lewis) [reply selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java:49]: Ah right. I think that might have been a holdover from when I was accidentally using autofuzz

File: selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java (hunk 2 above)
Comment (br-lewis) [reply selffuzz/src/test/java/com/code_intelligence/selffuzz/mutation/mutator/proto/ProtobufMutatorFuzzTest.java:73]:
```
Exception in thread "main" java.lang.IllegalStateException: Did not expect a finding, but got a stack trace:
== Java Exception: java.lang.IllegalArgumentException: Failed to execute method specified by @WithDefaultInstance("com.code_intelligence.selffuzz.mutation.mutator.proto.ProtobufMutatorFuzzTest#getDefaultInstance")
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.TypeLibrary.getDefaultInstance(TypeLibrary.java:174)
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.BuilderMutatorFactory.lambda$tryCreate$20(BuilderMutatorFactory.java:354)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.BuilderMutatorFactory.tryCreate(BuilderMutatorFactory.java:350)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.ChainedMutatorFactory.lambda$tryCreate$0(ChainedMutatorFactory.java:46)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at com.code_intelligence.selffuzz.jazzer.mutation.support.StreamSupport.findFirstPresent(StreamSupport.java:44)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.ChainedMutatorFactory.tryCreate(ChainedMutatorFactory.java:46)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.MutatorFactory.tryCreate(MutatorFactory.java:66)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.MutatorFactory.tryCreateInPlace(MutatorFactory.java:57)
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.MessageMutatorFactory.lambda$tryCreate$2(MessageMutatorFactory.java:46)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.MessageMutatorFactory.tryCreate(MessageMutatorFactory.java:43)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.ChainedMutatorFactory.lambda$tryCreate$0(ChainedMutatorFactory.java:46)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at com.code_intelligence.selffuzz.jazzer.mutation.support.StreamSupport.findFirstPresent(StreamSupport.java:44)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.ChainedMutatorFactory.tryCreate(ChainedMutatorFactory.java:46)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.MutatorFactory.tryCreate(MutatorFactory.java:66)
	at com.code_intelligence.selffuzz.jazzer.mutation.api.MutatorFactory.createOrThrow(MutatorFactory.java:41)
	at com.code_intelligence.selffuzz.mutation.mutator.proto.ProtobufMutatorFuzzTest.protobufMutatorTest(ProtobufMutatorFuzzTest.java:62)
	at jdk.internal.reflect.GeneratedMethodAccessor54.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.code_intelligence.selffuzz.jazzer.mutation.mutator.proto.TypeLibrary.getDefaultInstance(TypeLibrary.java:171)
	at com.code_intelligence.selffuzz.mutation.mutator.proto.ProtobufMutatorFuzzTest.getDefaultInstance(ProtobufMutatorFuzzTest.java:93)
	at com.google.protobuf.Descriptors$FieldDescriptor.<init>(Descriptors.java:1544)
	at com.google.protobuf.Descriptors$FieldDescriptor.<init>(Descriptors.java:1066)
	at com.google.protobuf.Descriptors$Descriptor.<init>(Descriptors.java:944)
	at com.google.protobuf.Descriptors$Descriptor.<init>(Descriptors.java:657)
	at com.google.protobuf.Descriptors$FileDescriptor.<init>(Descriptors.java:561)
	at com.google.protobuf.Descriptors$FileDescriptor.buildFrom(Descriptors.java:333)
	at com.google.protobuf.Descriptors$FileDescriptor.buildFrom(Descriptors.java:305)
	at com.code_intelligence.selffuzz.mutation.mutator.proto.ProtobufMutatorFuzzTest.getDefaultInstance(ProtobufMutatorFuzzTest.java:91)
	at com.code_intelligence.jazzer.tools.FuzzTargetTestWrapper.verifyFuzzerOutput(FuzzTargetTestWrapper.java:209)
```

**Review by fmeum — APPROVED** [review]

**Pull request merged**